library("optparse")

option_list = list(make_option("--windowFile", type = "character", default = NULL, help  = "Window file returned by makeWindows."),
				   make_option("--signalFile", type = "character", default = NULL, help = "Signal file generated by deepTools on regions defined in window_file."),
				   make_option("--BED", type = "character", default = NULL, help = "BED file of interest."),
				   make_option("--expression_file", type = "character", default = NULL, help = "Gene expression profile [1st: gene_id/gene_name; 2nd: log transformed value]."),				   
				   make_option("--min_N", type = "integer", default = 20, help = "Minimal number of genes to calculate summary scores [20]."),
				   make_option("--min_methylation", type = "numeric", default = 0, help = "Minimal methylation for further analysis [0]."),
				   make_option("--max_methylation", type = "numeric", default = 1, help = "Maximal methylation for further analysis [1]."),				   			   
				   make_option("--method", type = "character", default = 'Spearman', help = "Summary method [Pearson, Spearman, median, mean]."),
				   make_option("--outTag", type = "character", default = NULL, help = "Prefix for output files."))		   
args <- parse_args(OptionParser(option_list=option_list))

# args = list()
# args$windowFile = 'hg19_windows.bed'
# args$signalFile = 'debug_score.tsv'
# args$outTag = 'ff'
# args$BED = 'esophagus_T_MHB.txt'
# args$expression_file = 'GSM4505886_T1.txt'
# args$method = 'median'
# args$min_N = 20

# check parameters
if(is.null(args$windowFile))
	stop("\nArgument windowFile is required.\n")
if(is.null(args$signalFile))
	stop("Argument signalFile is required.\n")
if(is.null(args$BED))
	stop("Argument BED is required.\n")
if(is.null(args$expression_file))
	stop("Argument expression_file is required.\n")
if(is.null(args$outTag))
	stop("Argument outTag is required.\n")
if(!args$method %in% c('Pearson', 'Spearman', 'mean', 'median'))
	stop("Argument method should be one of Pearson, Spearman, mean and median.\n")

# load required libraries
library("GenomicRanges")

# read window and signal file
tr = read.table(file = args$windowFile, row.names = NULL, header = FALSE, comment.char = "#")
ts = read.table(file = args$signalFile, row.names = NULL, header = FALSE, comment.char = "#")
gr = GRanges(seqnames = as.character(tr[,1]), 
			ranges = IRanges(start = as.numeric(tr[,2]), end = as.numeric(tr[,3])),
			name = as.character(tr[,4]))
gs = GRanges(seqnames = as.character(ts[,1]), 
			ranges = IRanges(start = as.numeric(ts[,2]), end = as.numeric(ts[,3])),
			score = as.numeric(ts[,4]))

# after sorting, two GRanges should order in the same way
gr = sort(gr)
gs = sort(gs)
gr$score = gs$score

# update gene_id, gene_name, window
chunks = strsplit(gr$name, "_")
gr$gene_id = sapply(chunks, function(z) z[1])
gr$gene_name = sapply(chunks, function(z) z[2])
gr$window = sapply(chunks, function(z) paste0(z[3], '_', z[4]))
gr = gr[!is.na(gr$score)]

# update Feature
tb = read.table(file = args$BED, row.names = NULL, header = FALSE, comment.char = "#")
gb = GRanges(seqnames = as.character(tb[,1]), ranges = IRanges(start = as.numeric(tb[,2]), end = as.numeric(tb[,3])))
gr$Feature = countOverlaps(gr, gb, type = 'any', ignore.strand = TRUE) > 0

# preprocess gene expression file
tmp = read.table(file = args$expression_file, row.names = NULL, header = TRUE, sep = '\t')
tmp = tmp[!is.na(tmp[,2]), , drop = FALSE]
agg = aggregate(as.numeric(tmp[,2]), by = list(factor(tmp[,1])), FUN='mean')
ex = as.numeric(agg[,2])
names(ex) = as.character(agg[,1])

# process each window
sharedGenes = intersect(names(ex), gr$gene_name)
gr = gr[gr$gene_name %in% sharedGenes]

# filter methylation
gr = gr[(gr$score > args$min_methylation) & (gr$score < args$max_methylation)]

gf = gr[gr$Feature]
wid = unique(gr$window)
out = matrix(NA, nrow = length(wid), ncol = 4, dimnames = list(wid, c('Num_All', 'Num_Feature', 'Score_All', 'Score_Feature')))

for(i in 1:length(wid)){
	window = wid[i]
	g1 = gr[gr$window == window]
	g2 = gf[gf$window == window]
	out[i, 1] = length(g1)
	out[i, 2] = length(g2)
	m1 = g1$score
	e1 = ex[g1$gene_name]
	m2 = g2$score
	e2 = ex[g2$gene_name]
	if(args$method %in% c('Pearson', 'Spearman')){
		out[i, 3] = cor(m1, e1, method = tolower(args$method))
		if(length(g2) > args$min_N){
			out[i, 4] = cor(m2, e2, method = tolower(args$method))
		}
	} else if(args$method == 'mean') {
			out[i, 3] = mean(e1)
			out[i, 4] = mean(e2)
	} else if(args$method == 'median'){
			out[i, 3] = median(e1)
			out[i, 4] = median(e2)
	}
}
write.table(out, file = paste0(args$outTag, '_RA.tsv'), row.names = TRUE, col.names = TRUE, sep = '\t', quote = FALSE)
